assets/model_monitoring/components/tests/unit/test_numerical_data_drift_metrics.py

```diff
@@ -1,87 +1,342 @@
+        y_array_extended = np.append(y_array,[min_value, max_value])    
+
+        #select the amount of bins used. The smaller amount of bins is recommended to ensure both x_array and y_array have well populated histograms 
+        if min_value == max_value:
+            number_of_bins = 1
+            bin_edges = np.array([min_value/1.01, min_value*1.01])
+    
+        else:
+            #compute optimal bin edges for both distributions after having been extended to [min_value,max_value]
+            x_bin_edges=np.histogram_bin_edges(x_array_extended, bins='sturges')
+            y_bin_edges=np.histogram_bin_edges(y_array_extended, bins='sturges')
+            number_of_bins = min(len(x_bin_edges), len(y_bin_edges))
+            bin_edges=np.linspace(min_value, max_value, number_of_bins)
+
+        return bin_edges
 
-        x = np.random.normal(x_mean, std_dev, x_n_obs)
 
-        x_pd_df = pd.DataFrame(data=x, columns=column_values)
-        y_pd_df = pd.DataFrame(data=x, columns=column_values)
+    # # # EXPECTED NORMALIZED WASSERSTEIN DISTANCE # # # 
 
-        spark = init_spark()
+    def normed_wasserstein_distance_numerical(self, x_array, y_array):
+        norm = max(np.std(x_array),0.001)
+        return wasserstein_distance(x_array, y_array) / norm
 
-        x_df = spark.createDataFrame(x_pd_df)
-        y_df = spark.createDataFrame(y_pd_df)
+    # # # EXPECTED JENSEN-SHANNON DISTANCE # # # 
 
-        output_df = compute_numerical_data_drift_measures_tests(
-                x_df,
-                y_df,
-                x_df.count(),
-                y_df.count(),
-                "NormalizedWassersteinDistance",
-                column_values,
-                numerical_threshold)
+    def jensen_shannon_distance_numerical(self, x_array, y_array):
+        bin_edges = self.compute_optimal_histogram_bin_edges(x_array, y_array)
+        x_percent = np.histogram(x_array, bins=bin_edges)[0] / len(x_array)
+        y_percent = np.histogram(y_array, bins=bin_edges)[0] / len(y_array)
 
-        metric_value = self.get_metric_value(output_df, "NormalizedWassersteinDistance")
-        self.assertAlmostEqual(0.0, metric_value, 4)
+        return distance.jensenshannon(x_percent, y_percent, base=2)
 
-    def test_compute_numerical_data_drift_metrics_normalized_wasserstein_distance(self):
-        """Test compute normalized wasserstein distance for numerical metrics."""
-        std_dev = 15
+    # # # EXPECTED PSI # # #
+    
+    def psi_numerical(self, x_array, y_array):
+        bin_edges = self.compute_optimal_histogram_bin_edges(x_array, y_array)
+
+        x_count = np.histogram(x_array, bins=bin_edges)[0]
+        y_count = np.histogram(y_array, bins=bin_edges)[0] 
+        
+        # Laplace smoothing (incrementing the count of each bin by 1)
+        # to avoid zero values in bins and have the SPI value be finit
+        x_count = x_count + 1
+        y_count = y_count + 1 
+
+        x_percent = x_count / len(x_array) 
```

Comment:
Do we need to increment the denominator (like `len(x_array)` too so that the percentage numbers add up to 100?

Reply:
Yes! Great Catch! I did not see the issue as it was both present in the source code as well as the test, so all tests passed! 
Here is a snippet which reproduces the issue:
```python
x_array = [1, 2, 2, 3, 3, 3]
bin_edges = [0.5, 1.5, 2.5, 3.5, 4.5]

x_count = np.histogram(x_array, bins=bin_edges)[0]
print(f'x_count before: {x_count}')

x_count = x_count + 1
print(f'x_count after: {x_count}')

x_percent = x_count / len(x_array)
print(f'x_percent: {x_percent}')
print(f'Sum of percentages: {sum(x_percent)}')
```
```
x_count before: [1 2 3 0]
x_count after: [2 3 4 1]
x_percent: [0.33333333 0.5        0.66666667 0.16666667]
Sum of percentages: 1.6666666666666667
```

After applying fix: 
```python
x_array = [1, 2, 2, 3, 3, 3]
bin_edges = [0.5, 1.5, 2.5, 3.5, 4.5]

x_count = np.histogram(x_array, bins=bin_edges)[0]
print(f'x_count before: {x_count}')

x_count = x_count + 1
print(f'x_count after: {x_count}')

x_percent = x_count / (len(x_array) + len(x_count))
print(f'x_percent: {x_percent}')
print(f'Sum of percentages: {sum(x_percent)}')
```
```
x_count before: [1 2 3 0]
x_count after: [2 3 4 1]
x_percent: [0.2 0.3 0.4 0.1]
Sum of percentages: 1.0
```
I fixed the issue both in the test and source code.